File: jina_handler.py
import httpx
import os

async def scrape_with_jina(url: str, api_key: str, timeout_seconds: int) -> str:
    """
    Scrapes the given URL using the Jina.ai API.
    """
    jina_url = f"https://r.jina.ai/{url}"
    async with httpx.AsyncClient() as client:
        headers = {
            "Authorization": f"Bearer {api_key}"
        }
        response = await client.get(jina_url, headers=headers, follow_redirects=True, timeout=float(timeout_seconds))
        response.raise_for_status()
        return response.text

File: Api/Services/JinaHandler.cs
using System.Net.Http.Headers;
using System.Text.Json;

namespace JinaFirecrawlApi.Services;

public class JinaHandler : IJinaHandler
{
    private readonly IHttpClientFactory _httpClientFactory;

    public JinaHandler(IHttpClientFactory httpClientFactory)
    {
        _httpClientFactory = httpClientFactory;
    }

    public async Task<string> ScrapeWithJina(string url, string authHeader, int timeoutSeconds)
    {
        var httpClient = _httpClientFactory.CreateClient(nameof(JinaHandler));
        httpClient.Timeout = TimeSpan.FromSeconds(timeoutSeconds);
        httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", authHeader.Replace("Bearer ", ""));
        var fallbackTimeout = timeoutSeconds - 4;
        if (fallbackTimeout > 0) httpClient.DefaultRequestHeaders.Add("X-Timeout", fallbackTimeout.ToString());

        var payload = new { url = url };
        var content = new StringContent(JsonSerializer.Serialize(payload), System.Text.Encoding.UTF8, "application/json");

        using var response = await httpClient.PostAsync("https://r.jina.ai", content);

        response.EnsureSuccessStatusCode();

        var responseString = await response.Content.ReadAsStringAsync();
        return responseString;
    }
}
